### PR TITLE
refactor(transparent-proxy): move executables to config

### PR DESCRIPTION
**Pull request by bartsmykla**
> [!WARNING]
> I suggest doing review on commit per commit base. I new it'll be a bigger PR so I did my best to commit changes in small chunks.

## Core changes

- Centralized iptables executable management and verification

  - Introduced `Executables` structures to simplify locating and verifying executables for various iptables versions (nft, legacy) and IPv4/IPv6.
  - Eliminated hardcoded executable names, replacing them with retrieved paths from the `Executables` structures.
  - Ensured usage of correct executables (`iptables`, `iptables-restore`, `iptables-save`, etc.) throughout the codebase.
  - During initialization, verifies if the discovered iptables binary meets minimum requirements (presence of "nat" and "raw" tables, necessary modules etc.).

- Removed redundant `iptables` call for conntrack zone check

  - Eliminated the unnecessary iptables call to verify conntrack zone split after configuration initialization.
  - Leveraged the single, pre-initialized iptables executable for all interactions.
  - Streamlined rules builder logic by removing the need to pass the iptables path during construction (previously required a workaround during dry runs).

- Handle legacy iptables versions for mode verification

  Previously, iptables mode verification relied on the `--version` flag. However, older iptables versions (e.g., v1.6.1 in Ubuntu 18.4 or v1.4.21 in CentOS 7) lack this functionality.

  Improved mode verification for these legacy versions by:

  - Handling cases where the `--version` flag is unavailable for `iptables-save` and `iptables-restore`.
  - Identifying legacy mode based on the absence of a "mode" indicator (like "nf_tables") in the version output (e.g., `iptables v1.6.1`).

- Improved handling of dropping invalid packets:

  - This commit introduces verification of the mangle table presence within the `Executables.Functionality.Tables` configuration.
  - The `ShouldDropInvalidPackets` method is updated to consider the mangle table presence for the appropriate IP version (IPv4 or IPv6) when determining if dropping invalid packets is enabled.

- Enhanced Initialization for IPvX Executables and Add IPv6 Outbound Address

  This change enhances the initialization process for IPv4 and IPv6 executables within the ExecutablesIPvX struct and adds a function to configure a dedicated IPv6 outbound address. The changes include:

  Adding `configureIPv6OutboundAddress` function:
  - Sets up a dedicated IPv6 address (::6) on the loopback interface (lo) for transparent proxy functionality.
  - Ensures the transparent proxy can identify and process IPv6 traffic correctly, avoiding redirection loops.
  - Uses the netlink library to add the IPv6 address.
  - Handles cases where the address might already exist to avoid unnecessary errors.

- Refactored `Executable` struct to use prefix instead of ipv6 property

  The `prefix` property in the Executable struct now represents the prefix used for iptables executables, distinguishing between IPv4 and IPv6 operations. For IPv4, it's `iptables` (covering binaries like `iptables`, `iptables-restore`, or `iptables-save`). For IPv6, it's `ip6tables` (covering binaries like `ip6tables`, `ip6tables-restore`, or `ip6tables-save`). This refactoring enhances clarity and utility by unifying how executable names are constructed, temporary file names for iptables rules are generated, and other related operations are performed.

- Enhanced `InitializedExecutable` to support default arguments

  The `args` field in `InitializedExecutable` now stores a set of default arguments or flags that are consistently included in every execution of the executable. These default arguments are prepended to any additional arguments passed in the Exec method. This enhancement ensures that specific flags or options are consistently applied whenever the executable is invoked.

- Enhanced `Executables` initialization with default args for `iptables-restore`

  Refactored `Executables`' Initialize method to incorporate default arguments for iptables-restore using `buildRestoreArgs`. This update ensures that the IptablesRestore executable component is initialized with appropriate default flags based on the provided Config.

  `buildRestoreArgs` constructs a slice of flags for restoring iptables rules based on the provided wait time, wait interval, and iptables mode.

  This function generates a list of command-line flags to be used with `iptables-restore`, tailored to the given parameters:

  - By default, it includes the `--noflush` flag to prevent flushing of existing rules.
  - If the iptables mode is not legacy, it returns only the `--noflush` flag.
  - For legacy mode, it conditionally adds the `--wait` and `--wait-interval` flags based on the provided wait time and waitInterval.

- Refined `RetryConfig` struct

  This change streamlines the `RetryConfig` struct by making the following adjustments:
  - The `MaxRetries` field has been changed from a pointer to a direct integer type. This change is based on the determination that distinguishing whether `MaxRetries` was explicitly set is unnecessary. Using a direct integer simplifies the logic by avoiding the need to check for nil values and ensures that the default behavior is straightforward with zero indicating no retries.
  - Comments have been added to describe the `MaxRetries` and `SleepBetweenRetries` properties.

- Introduced logging abstraction and fixed conntrack rules

  - Implemented a logging abstraction layer within `InitializedConfig` for improved transparency during configuration and iptables operations.
  - Removed unnecessary logging related to successful transparent proxy setup in dry run mode.
  - Fixed the generation of conntrack zone split rules in dry run mode.

- Enhanced `InitializedExecutables` with rule writing and restore functions

  This change introduces enhancements to the `InitializedExecutables` struct by adding new function: `writeRulesToFile` and a method `Restore`. These functions streamline the process of managing iptables rules, from writing to temporary file to restoring them, with retry mechanisms.

  - `writeRulesToFile` function: writes iptables rules to a temporary file. This method logs the file path and contents for transparency.
  - `Restore` method: executes `iptables-restore` with the rules from the temporary file created by `writeRulesToFile`. This method implements retry logic defined in the newly added retry property of `InitializedExecutables`, attempting the restore operation up to the specified number of retries and logging each attempt.

- Added few comments to outbound rules generation for NAT table

- Replaced old iptables restoration logic with the new one

  The new restoration logic within `InitializedConfig` effectively replaces the old iptables restorer abstraction. As part of this update, all obsolete structures related to the old iptables restoration logic have been removed.

  The `ShouldFallbackDNSToUpstreamChain` method has been removed from `InitializedConfig`. This decision was made due to its limited use, being utilized in only one place within the project.

  Config initialization now runs `iptables` executables under the hood, so it needs to be run in appropriate network namespace. This was adjusted for CNI.

- Enhanced Dry Run to generate both IPv4 and IPv6 proxy rules

  This change improves the dry run functionality of transparent proxy installation by simultaneously generating iptables rules for both IPv4 and IPv6 when required. Previously, rule generation was restricted to either IPv4 or IPv6 based on the IPv6 property in the `InitializedConfig`. This update ensures comprehensive rule validation for setups supporting dual-stack configurations.

- Added Ubuntu 18.04 and RHEL 8 to transparent proxy tests

## Small changes and improvements

- Added comment to Log jump target with explanation of log levels

  Descriptions of jump target taken from iptables manual (https://linux.die.net/man/8/iptables). Log levels taken from: https://git.netfilter.org/iptables/tree/extensions/libebt_log.c#n27

- Moved `DebugLogLevel` const to consts package + add comment

- Renamed `DebugLogLevel` to `LogLevelDebug` for consistency

  If somehow in the future we'll decide to add another log levels its better to have a common part in the consts name at the beginning

- Moved `getLoopbackInterfaceName` helper func to the bottom

  To be consistent I decided to always place helper functions at the bottom of files, so it should be easier to read/review the code in the future

- Added comment to the `getLoopbackInterfaceName` helper function

- Refactored flag-related constants for clarity and efficiency

- Renamed `Flags` map to `FlagsShortLongMap` to accurately reflect its purpose of mapping short and long flag names.

- Removed unused fields from `FlagsShortLongMap` to improve code maintainability

- Renamed `config` parameter name to `cfg` to be consistent

- Replaced `errors.Errorf` with `errors.Wrapf` in few places

- Improved documentation and error handling in `BuildIPTablesForRestore` func

- Refactored `Flag` Lookup with `FlagVariationsMap`

  - Renamed `FlagsShortLongMap` to the more descriptive `FlagVariationsMap`.
  - Removed unused type aliases `Long` and `Short`.

- Improved error handling and update description in `execCmd` function

  - Enhanced the formatting of the stderr output when an error occurs. The stderr output is now trimmed of whitespace, split into lines, and joined with commas to provide a more concise and readable error message.
  - Updated the error message to use errors.Errorf for consistent error formatting with the inclusion of the original error and formatted stderr content.

- Introduced `IptablesCommandByFamily` map for generic IPv4/IPv6 handling

  - Created a new `IptablesCommandByFamily` map to map a boolean value representing IPv4/IPv6 usage to the corresponding iptables command name.
  - This enables writing generic code that works for both IPv4 and IPv6 scenarios without code duplication.

- Removed obsolete parameter abstractions since `iptables-restore` parameters are now constructed within the `buildRestoreArgs` function in `config_executables.go`. This change eliminates unnecessary complexity in parameter handling.

### Checklist prior to review

- [x] [Link to relevant issue][1] as well as docs and UI issues
  - No relevant issues
- [x] This will not break child repos: it doesn't hardcode values (.e.g "kumahq" as a image registry) and it will work on Windows, system specific functions like `syscall.Mkfifo` have equivalent implementation on the other OS
  - It won't
- [x] Tests (Unit test, E2E tests, manual test on universal and k8s)
  - Don't forget `ci/` labels to run additional/fewer tests
- [x] Do you need to update [`UPGRADE.md`](../blob/master/UPGRADE.md)?
- [x] Does it need to be backported according to the [backporting policy](../blob/master/CONTRIBUTING.md#backporting)? ([this](https://github.com/kumahq/kuma/actions/workflows/auto-backport.yaml) GH action will add "backport" label based on these [file globs](https://github.com/kumahq/kuma/blob/master/.github/workflows/auto-backport.yaml#L6), if you want to prevent it from adding the "backport" label use [no-backport-autolabel](https://github.com/kumahq/kuma/blob/master/.github/workflows/auto-backport.yaml#L8) label) --

[1]: https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword

### CODE REVIEW

**Review by slonka — COMMENTED** [review]
Left a couple of comments, nothing alarming - if we don't come up with any changes after answering them I'll 👍

File: pkg/transparentproxy/config/config_executables.go [hunk 1]
```diff
@@ -0,0 +1,703 @@
...
+	if err != nil {
+		return isVersionMissing(err.Error()) && mode == IptablesModeLegacy
+	}
+
+	if stderr != nil && stderr.Len() > 0 && isVersionMissing(stderr.String()) {
+		return mode == IptablesModeLegacy
+	}
+
+	matched := IptablesModeRegex.FindStringSubmatch(stdout.String())
+	if len(matched) == 2 {
```
[the remark this answers is not in the document]
Comment (bartsmykla) [reply pkg/transparentproxy/config/config_executables.go:521]: That's correct, but `FindStringSubmatch` when submatch is found returns a slice with first item be full match from regex:
```go
package main

import (
	"fmt"
	"regexp"
)

func main() {
	matches := regexp.MustCompile(`(?m)^ip6?tables(?:.*?\((.*?)\))?`).
		FindStringSubmatch("iptables v1.8.9 (nf_tables)\n")
	fmt.Println(len(matches), matches)

	matches2 := regexp.MustCompile(`(?m)^ip6?tables(?:.*?\((.*?)\))?`).
		FindStringSubmatch("iptables v1.6.1")
	fmt.Println(len(matches2), matches2)
}
```
will return
```
2 ["iptables v1.8.9 (nf_tables)", "nf_tables"]
2 ["iptables", ""]
```

File: pkg/transparentproxy/config/config_executables_functionality.go [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+
+	return nil
+}
+
+func verifyTable(
+	ctx context.Context,
+	iptablesSave InitializedExecutable,
+	table TableName,
+) (bool, string) {
+	stdout, _, err := iptablesSave.Exec(ctx, FlagTable, string(table))
```
[the remark this answers is not in the document]
Comment (bartsmykla) [reply pkg/transparentproxy/config/config_executables_functionality.go:94]: `iptables-save` only generates a content of file which then can be used to run `iptables-restore`, which actually restores the rules

File: test/transparentproxy/config.go [hunk 3]
```diff
@@ -61,18 +61,15 @@ var defaultConfig = TransparentProxyConfig{
 		"Ubuntu 24.04":      "ubuntu:noble-20240605",
 		"Ubuntu 22.04":      "ubuntu:jammy-20240530",
 		"Ubuntu 20.04":      "ubuntu:focal-20240530",
+		"Ubuntu 18.04":      "ubuntu:bionic-20230530",
 		"Debian 12":         "debian:bookworm-20240612",
 		"Debian 11":         "debian:bullseye-20240612",
 		"Debian 10":         "debian:buster-20240612",
 		"RHEL 9":            "redhat/ubi9:9.4-1123",
+		"RHEL 8":            "redhat/ubi8:8.10-901.1717584420",
```
[the remark this answers is not in the document]
Comment (bartsmykla) [reply test/transparentproxy/config.go:69]: I was looking for `cos` docker images, but without luck, so I don't think we'll be able to test them.

I don't mind adding fedora/archlinux, but I would prefer to do it in the next PR, as this is already quite big and we are suffering from the huge CI congestion.

File: pkg/transparentproxy/iptables/consts/consts.go [hunk 4]
```diff
@@ -35,53 +42,92 @@ const (
...
-	ChainPostrouting = "POSTROUTING"
+	ChainPrerouting   = "PREROUTING"
+	ChainInput        = "INPUT"
+	ChainForward      = "FORWARD"
+	ChainOutput       = "OUTPUT"
+	ChainPostrouting  = "POSTROUTING"
+	ChainDockerOutput = "DOCKER_OUTPUT"
 )
 
+// DockerOutputChainRegex is aregular expression used to identify the presence
```
[the remark this answers is not in the document]
Comment (bartsmykla) [reply pkg/transparentproxy/iptables/consts/consts.go]: addressed

File: pkg/transparentproxy/config/config_executables.go [hunk 5]
```diff
@@ -0,0 +1,703 @@
...
+	}
+	initialized.IPv4 = ipv4
+
+	ipv6, ipv6Err := c.IPv6.Initialize(ctx, l, cfg)
+	if ipv6Err != nil {
+		errs = append(errs, ipv6Err)
+	}
+	initialized.IPv6 = ipv6
+
+	if len(errs) == 2 {
```
[the remark this answers is not in the document]
Comment (bartsmykla) [reply pkg/transparentproxy/config/config_executables.go]: I had assumption that you could want to configure tproxy only IPv6, but as there are still few missing bits to allow it a decided to change the logic to always fail if initialization for IPv4 won't be successful.

File: test/transparentproxy/config.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (bartsmykla) [reply test/transparentproxy/config.go:69]: I added `fedora`

**Review by slonka — APPROVED** [review]
Can we run blackbox tests at least once on CI to get check it works? AFAIK we're not running them now, right?

**Comment by bartsmykla** [comment]
> Can we run blackbox tests at least once on CI to get check it works? AFAIK we're not running them now, right?

We don't and they are probably not in a great state, but to be honest I think they'll be replaced by transparent-proxy tests fully

**Comment by bartsmykla** [comment]
Ok, did final tests on GKE/COS, all looks good

**Pull request merged**